core/validation/validate.py

```diff
@@ -335,14 +336,18 @@ def validate_nullable(
    """
     for column in sheet.columns:
         if column in non_nullable and (pd.isna(sheet[column]).any() | (sheet[column].astype("string") == "").any()):
-            failed_indexes = sheet.index[(pd.isna(sheet[column]) | (sheet[column].astype("string") == ""))]
-
-            non_nullable_constraint_failure.append(
-                vf.NonNullableConstraintFailure(
-                    sheet=sheet_name,
-                    column=column,
-                    row_indexes=failed_indexes.tolist(),
-                )
+            non_nullable_constraint_failure.extend(
+                [
+                    vf.NonNullableConstraintFailure(
+                        sheet=sheet_name,
+                        column=column,
+                        row_indexes=[index],
+                        failed_row=pd.DataFrame(dict(row), index=[index]),
+                    )
+                    for index, row in sheet[
+                        (pd.isna(sheet[column]) | (sheet[column].astype("string") == ""))
+                    ].iterrows()
+                ]
```

Comment:
I think this may be more clear to split it into two lines: 

Step 1: Filter dataframe to offending rows

Step 2: iterate the rows, returning a Failure for each

Can we also just return the row as a series instead of creating a df with a single row?